### PR TITLE
Add code and contract getters to CosmWasmClient

### PR DESCRIPTION
**Pull request by webmaster128**
This is an alternative to #112, with some differences:
* use our own types `Code`, `CodeDetails`, `Contract` exposed from @cosmwasm/sdk that don't need to match the REST response directly
* Prepare for https://github.com/cosmwasm/wasmd/issues/75, avoiding the `/wasm/contract` getter, which has no long term usage
* Try to come up with the best possible names instead of mirroring REST backend or client names

### CODE REVIEW

File: packages/sdk/src/cosmwasmclient.ts [hunk 1]
```diff
@@ -166,6 +190,42 @@ export class CosmWasmClient {
...
+  }
+
+  public async getCodeDetails(codeId: number): Promise<CodeDetails> {
+    const result = await this.restClient.getCode(codeId);
+    return {
+      wasm: result,
+    };
+  }
+
+  public async getContracts(codeId: number): Promise<readonly Contract[]> {
```
Comment (ethanfrey) [inline packages/sdk/src/cosmwasmclient.ts]: Fine for current rest api, but we do want to add the contract address to all of these soon. I guess it is blocked on wasmd changes, but very important for usa ility.

For the name app, I added address to the api, pulling the argument from get contact and just brute force downloading all contracts and filtering locally for get contracts. This gives me the proper api now, and I can clean up the implementation once wasmd is enhanced.

Comment (webmaster128) [reply packages/sdk/src/cosmwasmclient.ts]: Sounds like a good approach and I agree we want the address here ASAP (make a comment in the `Contract` interface already). Can you show me that code? I don't know how address generation works for contract instances and find it hard to follow your explanation.

**Comment by webmaster128** [comment]
> Also make the contract generic T over init_msg, default object, as the caller often has more info.

When I worked on this, I thought `init_msg` should not be part of the list at all (see https://github.com/cosmwasm/wasmd/issues/80)

File: packages/sdk/src/cosmwasmclient.ts (hunk 1 above)
Comment (webmaster128) [reply packages/sdk/src/cosmwasmclient.ts]: I found the code/implementation in name-app you're refering to. This is .... wow, creative. And it does the job 🎩. But even in our small short living testing world this would easily send 100 requests to get the addresses of a single `codeId`. If there is no very good reason we need the interface now, I'd wait until the backend is ready.
